### PR TITLE
Use OpaquePointer as a safe support for region based allocators

### PR DESCRIPTION
**Pull request by edi33416**

### CODE REVIEW

File: std/experimental/allocator/building_blocks/free_list.d [hunk 1]
```diff
@@ -532,7 +565,9 @@ struct ContiguousFreeList(ParentAllocator,
     {
         static if (maxSize == chooseAtRuntime) fl.max = max;
         static if (minSize == chooseAtRuntime) fl.min = max;
-        initialize(cast(ubyte[])(parent.allocate(bytes)), max);
+        auto nb = (bytes + OpaquePointer.sizeof - 1) / OpaquePointer.sizeof;
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/free_list.d]: This seems uncovered, take a look?

File: std/experimental/allocator/building_blocks/kernighan_ritchie.d [hunk 2]
```diff
@@ -333,12 +336,37 @@ struct KRRegion(ParentAllocator = NullAllocator)
             b.ptr, b.length);
     }
 
+    /// ditto
+    this(OpaquePointer[] b)
+    {
+        if (b.length * OpaquePointer.sizeof < Node.sizeof)
+        {
+            // Init as empty
+            assert(root is null);
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/kernighan_ritchie.d]: Uncovered

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 3]
```diff
@@ -142,17 +143,49 @@ struct BitmappedBlock(size_t theBlockSize, uint theAlignment = platformAlignment
...
     an appropriate contiguous hunk of memory. Regardless of the constructor
-    used, the destructor releases the memory by using $(D
-    ParentAllocator.deallocate).)
+    used, the destructor releases the memory by using `ParentAllocator.deallocate`.)
     )
     */
+    this(OpaquePointer[] data)
+    {
+        immutable a = data.ptr.effectiveAlignment;
+        assert(a >= size_t.alignof || !data.ptr,
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: No need for this assert, so you can delete it and also the line that defines `a`.

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 4]
```diff
@@ -142,17 +143,49 @@ struct BitmappedBlock(size_t theBlockSize, uint theAlignment = platformAlignment
...
+        immutable ulong bitsPerBlock = blockSize * 8 + 1;
+        // Get a first estimate
+        import std.conv : to;
+        _blocks = to!uint(totalBits / bitsPerBlock);
+
+        // Reality is a bit more complicated, iterate until a good number of
+        // blocks found.
+        for (; _blocks; --_blocks)
+        {
+            immutable controlWords = _blocks.divideRoundUp(64);
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: For clarity 64 could be replaced with `ulong.sizeof * 8`

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 5]
```diff
@@ -142,17 +143,49 @@ struct BitmappedBlock(size_t theBlockSize, uint theAlignment = platformAlignment
...
+            auto payload = data[controlWords * 8 .. $].roundStartToMultipleOf(
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: BUG: should be

```d
static assert(8 % OpaquePointer.sizeof == 0, "Unsupported pointer size");
auto payload = data[controlWords * 8 / OpaquePointer.sizeof .. $].roundStartToMultipleOf(alignment);
```

File: std/experimental/allocator/building_blocks/bitmapped_block.d [hunk 6]
```diff
@@ -142,17 +143,49 @@ struct BitmappedBlock(size_t theBlockSize, uint theAlignment = platformAlignment
...
+            _control[] = 0;
+            _payload = payload;
+            break;
+        }
+    }
+
+    /// Ditto
     this(ubyte[] data)
     {
         immutable a = data.ptr.effectiveAlignment;
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/bitmapped_block.d]: There's a fair amount of duplication with the other ctor, I think this should do some checks and adjustments and then forward to the OpaquePointer ctor.

File: std/experimental/allocator/building_blocks/free_list.d [hunk 7]
```diff
@@ -514,14 +543,18 @@ struct ContiguousFreeList(ParentAllocator,
     static if (!stateSize!ParentAllocator)
     this(size_t bytes)
     {
-        initialize(cast(ubyte[])(ParentAllocator.instance.allocate(bytes)));
+        auto nb = (bytes + OpaquePointer.sizeof - 1) / OpaquePointer.sizeof;
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/free_list.d]: Replace this idiom with `divideRoundUp` throughout.

File: std/experimental/allocator/building_blocks/region.d [hunk 8]
```diff
@@ -357,6 +379,24 @@ struct Region(ParentAllocator = NullAllocator,
...
+    foreach (T; AliasSeq!(ubyte, OpaquePointer))
+    {
+        // Create a 64 KB region
+        auto reg = Region!()(new T[1024 * 64 / T.sizeof]);
+        const b = reg.allocate(101);
+        assert(b.length == 101);
+        assert(reg.deallocateAll());
+        auto c = reg.allocateAll();
+        assert(c.length == 1024 * 64);
+        //// Destructor will free the memory
```
Comment (andralex) [inline std/experimental/allocator/building_blocks/region.d]: delete

File: std/experimental/allocator/common.d [hunk 9]
```diff
@@ -190,18 +190,17 @@ package size_t divideRoundUp(size_t a, size_t b)
 /**
 Returns `s` rounded up to a multiple of `base`.
 */
-@nogc nothrow pure
+@nogc nothrow pure @trusted
```
Comment (andralex) [inline std/experimental/allocator/common.d]: this cannot be trusted

Comment (andralex) [inline std/experimental/allocator/common.d]: you can make it trusted if it takes `ubyte[]`

Comment (aG0aep6G) [inline std/experimental/allocator/common.d]: > you can make it trusted if it takes ubyte[]

As far as I see, that doesn't affect safety. Slicing a `void[]` is safe, isn't it? You might get garbage when you cast it to something else, but that's always unsafe. dmd thinks it's safe.

But rounding up `s.ptr` without doing bounds checking seems unsafe. When rounding puts `p` beyond `s`'s bounds, then `end < p`, and `end - p` wraps around.

**Comment by andralex** [comment]
I'll force merge this.

**Comment by braddr** [comment]
I'm pretty sure this is the pull that's broken mainline and pull builds for windows 64 bit.

**Comment by wilzbach** [comment]
>> I'll force merge this.
> I'm pretty sure this is the pull that's broken mainline and pull builds for windows 64 bit.

-> #5451

**Comment by edi33416** [comment]
Looking at [auto tester log](https://auto-tester.puremagic.com/pull-history.ghtml?projectid=1&repoid=3&pullid=5403) for this PR, all the tests have passed on both the 6th and 7th of June.

This PR was previously failing because of a faulty dtor that I had added, but now is removed.

**Pull request merged**